Add FA2 and `sdpa` support for SigLIP

qubvel opened the pull request:
# What does this PR do?

Add flash attention 2 and sdpa (`torch.nn.functional.scaled_dot_product_attention`) attention implementations for SigLIP model.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #31138


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

qubvel committed:
Rebase to main

qubvel committed:
Fix attention implementation autoset for tex and vision configs

qubvel committed:
Fixup

qubvel committed:
Minor fixes

qubvel committed:
Fix copies

qubvel committed:
Fix attention_mask for FA2

qubvel committed:
Add eqvivalence tests for siglip

qubvel committed:
Remove right padding test

qubvel committed:
Uncomment flaky

qubvel committed:
Fix import

qubvel committed:
Add to docs

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31499). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel commented:
@amyeroberts there are some discrepancies with the attention mask, I am digging deeper to the equivalence testing

qubvel committed:
Fix test message

qubvel committed:
Add sdpa

qubvel committed:
Add sdpa equivalence test

qubvel committed:
Add siglip sdpa to docs

qubvel committed:
Fix typing for attention output

qubvel committed:
Add sdpa tests

qubvel committed:
Fix signature of FA2

qubvel committed:
Autoset attn_implementation in config

qubvel committed:
Rename bsz -> batch_size

qubvel committed:
Move back autoset attn method

qubvel committed:
Mark as flaky

qubvel committed:
Correct attention mask padding

qubvel committed:
[run-slow] siglip

qubvel commented:
Running test locally for flash attention and sdpa

```bash
RUN_SLOW=1 python -m pytest --verbose -m flash_attn_test \
    tests/models/siglip/test_modeling_siglip.py
```
<img width="1390" alt="Screenshot 2024-06-24 at 15 31 23" src="https://github.com/huggingface/transformers/assets/31920396/2b981867-fac0-4b0a-a9e4-dc73f50001e3">


```bash
RUN_SLOW=1 python -m pytest --verbose \
    tests/models/siglip/test_modeling_siglip.py -k "sdpa"
```

qubvel commented:
@amyeroberts @molbap please review if you have time

amyeroberts commented on src/transformers/models/siglip/modeling_siglip.py:
```diff
@@ -543,6 +825,33 @@ def _init_weights(self, module):
             module.weight.data.fill_(1.0)
 
+    @classmethod
+    def _autoset_attn_implementation(
```
c.f. https://github.com/huggingface/transformers/pull/31203#issuecomment-2145036503

I let this slip in for IDEFICS2, but it never should have been included

amyeroberts commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -225,6 +409,20 @@ def test_model_from_pretrained(self):
+    @require_torch_sdpa
+    @slow
+    @is_flaky()
+    def test_eager_matches_sdpa_inference(
+        self,
+        torch_dtype: str,
+    ):
```
nit 

```suggestion
    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
```

amyeroberts commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -382,6 +580,20 @@ def test_model_from_pretrained(self):
+    @require_torch_sdpa
+    @slow
+    @is_flaky()
+    def test_eager_matches_sdpa_inference(
+        self,
+        torch_dtype: str,
+    ):
```
nit 

```suggestion
    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
```

amyeroberts commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -587,6 +799,104 @@ def test_model_from_pretrained(self):
+    @require_torch_sdpa
+    @slow
+    @is_flaky()
+    def test_eager_matches_sdpa_inference(
+        self,
+        torch_dtype: str,
+    ):
```
```suggestion
    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
```

amyeroberts commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -646,6 +956,18 @@ def test_training_gradient_checkpointing_use_reentrant_false(self):
+    @require_torch_sdpa
+    @slow
+    @is_flaky()
+    def test_eager_matches_sdpa_inference(
+        self,
+        torch_dtype: str,
+    ):
```
```suggestion
    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
```

amyeroberts commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -55,6 +67,178 @@
     from transformers import SiglipProcessor
 
 
+class SiglipModelTesterMixin(ModelTesterMixin):
```
cc @ydshieh for comments / opinions on this mixin structure within the model's testing file

ydshieh replied:
IIRC, it's just to overwrite `test_eager_matches_sdpa_inference` (which is a large block) in `ModelTesterMixin)`.

So we don't really need to have this new class. However, there are 3 or more model test classes in this file. It's nice to have `SiglipModelTesterMixin` and just overwrite with mini block like

```
    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
        super().test_eager_matches_sdpa_inference(
            torch_dtype=torch_dtype,
            logit_keys=("pooler_output", "last_hidden_state"),
            use_attnetion_mask_options=(False,),
        )
```

amyeroberts commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -587,6 +799,104 @@ def test_model_from_pretrained(self):
+    @require_torch_gpu
+    @mark.flash_attn_test
+    @slow
+    @is_flaky()
```
Is it flaky for siglip? Do we know why? I know we have this decorator for the common tests, but not for the model-specific implementations

amyeroberts reviewed (commented):
Thanks for adding this! 

* Main comment is about the attention selection, we should instead be instantiating the model components with `from_config` (or possibly `_from_config`?) and passing in `attn_implementation=config.attn_implementation). 
* Could we extend this to CLIP and add both at the same time? 
* There should be a section added to the model doc page + benchmarks showing some times for improvements for SDPA and FA2 e.g. [like here for mistral](https://huggingface.co/docs/transformers/v4.41.3/en/model_doc/mistral#expected-speedups).

qubvel committed:
Add FA2 and sdpa docs

qubvel committed:
Style fix

qubvel committed:
Remove flaky for FA2 test

qubvel committed:
Change attention implementation set

qubvel commented:
@amyeroberts 

> Main comment is about the attention selection, we should instead be instantiating the model components with from_config (or possibly _from_config?) and passing in `attn_implementation=config.attn_implementation).

I tried to make it with `_from_config`, however, internal model components are not inherited from `PretrainedModel`, they are just pure `nn.Module` and don't have `_from_config` method. I changed the inheritance, but that led to other issues: models have to be included in docs and some fields have to be specified. 

Not sure we can change internal model components too, for example, change `SiglipVisionTransformer(nn.Module)` to  `SiglipVisionModel(SiglipPreTrainedModel)`. This will lead to incompatible checkpoints.

See implementation https://github.com/huggingface/transformers/pull/31499/commits/669c5378f4758b05345d6faa8fd51328ad57d32c

Do you have any thoughts on that?

>Could we extend this to CLIP and add both at the same time?

I hope we can, I will check this!

>There should be a section added to the model doc page + benchmarks showing some times for improvements for SDPA and FA2 e.g. [like here for mistral](https://huggingface.co/docs/transformers/v4.41.3/en/model_doc/mistral#expected-speedups).

Addressed in https://github.com/huggingface/transformers/pull/31499/commits/d41955d5ac12a25c3f64d445dab23668e6a2b34f

![speedup_plot](https://github.com/huggingface/transformers/assets/31920396/607c1bb8-eefa-45c4-94d9-6712dbd50aac)

sayakpaul commented:
Regarding the CLIP part: https://github.com/huggingface/transformers/pull/30390. Feel free to cherry-pick commits if that help. I got stuck in some FLAX tests that I never got time to get resolved.

ydshieh commented:
> I tried to make it with _from_config, however, internal model components are not inherited from PretrainedModel, they are just pure nn.Module and don't have _from_config method

Hi, regarding this part, even if an internal component is just `nn.Module`, we can still passing `config` to its `__init__`, just like what is done in `GemmaDecoderLayer`. Hope this could help.

```python
class GemmaDecoderLayer(nn.Module):
    def __init__(self, config: GemmaConfig, layer_idx: int):
        super().__init__()
        self.hidden_size = config.hidden_size

        self.self_attn = GEMMA_ATTENTION_CLASSES[config._attn_implementation](config=config, layer_idx=layer_idx)
```

qubvel committed:
Change attn_implementaiton propogation

qubvel commented:
I changed the attention implementation propagation as follows (commit https://github.com/huggingface/transformers/pull/31499/commits/23457a2233bf62cf3c3e7c24d6d0b08f07c78864):

1. Initialize the `*Model` class instead of the `*Transformer` module to utilize the `_from_config` method.
2. Use only the submodule of the `*Model` to maintain the overall model structure and weights for backward compatibility.

```python
# First, initialize the text and vision models with proper attention implementation
text_model = SiglipTextModel._from_config(text_config, attn_implementation=config._attn_implementation)
vision_model = SiglipVisionModel._from_config(vision_config, attn_implementation=config._attn_implementation)

# Second, get the text and vision submodules (for backward compatibility)
self.text_model = text_model.text_model
self.vision_model = vision_model.vision_model
```

With this approach, the underlying modules will exhibit the same behavior for the `attn_implementation` setting.

The disadvantage of this method is that the `post_init()` method is called twice: once for each `*Model` and again for the parent model.

@amyeroberts please let me know what you think.

molbap commented on src/transformers/models/siglip/modeling_siglip.py:
```diff
@@ -421,6 +442,266 @@ def forward(
+    SDPA API.
+    """
+
+    is_casual = False
```
_light hearted chuckle_
```suggestion
    is_causal = False
```

qubvel replied:
https://github.com/huggingface/transformers/pull/31499/commits/2647b9e94781a07b8742ed3163ac3839662e46ff

molbap commented on src/transformers/models/siglip/modeling_siglip.py:
```diff
@@ -503,6 +783,8 @@ class SiglipPreTrainedModel(PreTrainedModel):
         "SiglipEncoderLayer",
         "SiglipMultiheadAttentionPoolingHead",
     ]
+    _supports_flash_attn_2 = True
+    _supports_sdpa = True
```
:tada:

molbap reviewed (commented):
Did a small review, then off so posting it: mostly typos/minor suggestions :)

amyeroberts commented on docs/source/en/model_doc/siglip.md:
```diff
@@ -107,6 +107,88 @@ A list of official Hugging Face and community (indicated by 🌎) resources to h
+
+## Expected speedups
+
+Below is an expected speedup diagram that compares inference time between the native implementation in transformers using `google/siglip-so400m-patch14-384` checkpoint in `float16` precision and the Flash Attention 2 / SDPA version of the model using different batch sizes.
```
❤️

amyeroberts reviewed (approved):
🔥 🔥 🔥 🔥 🔥 🔥 

Thanks for adding and iterating on this! Re the double call to `post_init`, it's not ideal, but I think should be fine as on the second call all the layers should be marked as initialized

qubvel committed:
Fix typos

qubvel committed:
Add modality to assert message

qubvel commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -55,6 +67,178 @@
+                    with torch.backends.cuda.sdp_kernel(
+                        enable_flash=enable_kernels,
+                        enable_math=True,
+                        enable_mem_efficient=enable_kernels,
```
Copied from the original test, not sure I get the motivation of this check. Ideally, the equivalence and backend dispatching should be handled and tested by torch, isn't it?

qubvel replied:
But it can be changed to something like

```python
cases = [
    (use_mask, output_attentions, sdpa_backend, batch_size)
    for use_mask in use_attention_mask_options
    for output_attentions in [True, False]
    for sdpa_backend in [
        SDPBackend.FLASH_ATTENTION,
        SDPBackend.EFFICIENT_ATTENTION,
        [SDPBackend.FLASH_ATTENTION, SDPBackend.MATH],
        [SDPBackend.EFFICIENT_ATTENTION, SDPBackend.MATH],
    ]
    for batch_size in [1, 5]
]
```

to test different backends

qubvel replied:
https://github.com/huggingface/transformers/pull/31499/commits/26c6770c20ef6a3695946104f731f3952dd06dbf

qubvel committed:
Add more sdpa backends in test

qubvel commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -55,6 +67,178 @@
+                raise ValueError("The SDPA model should have SDPA attention layers")
+
+            # We use these for loops instead of parameterized.expand just for the interest of avoiding loading/saving the model each time,
+            # but it would be nicer to have an efficient way to use parameterized.expand
```
I just tried to rewrite with this approach
1) There will be too many sdpa test cases, ~200 cases for siglip models, and they are not much informative compared to what we have now
2) It seems like `@parameterized` replaces the original method `test_eager_matches_sdpa_inference` with 3 methods `test_eager_matches_sdpa_inference_[0, 1, 2]_float*` in the parent class making it hard to override original `test_eager_matches_sdpa_inference`. There might be situations when the original and overridden methods will be called, thus I will keep the signature identical to avoid this. This may also be a problem for future tests and we, probably, have to find a safe approach to override parameterized methods.

molbap replied:
alright, that's wel thought - the double test run would be bad indeed. LGTM

qubvel committed:
[run slow] siglip

qubvel commented on tests/models/siglip/test_modeling_siglip.py:
```diff
@@ -593,6 +799,100 @@ def test_model_from_pretrained(self):
+
+                self.assertTrue(
+                    torch.allclose(outputs.logits_per_image, outputs_fa.logits_per_image, atol=4e-2, rtol=4e-2),
+                    f"Logits max diff: {torch.max(torch.abs(outputs.logits_per_image - outputs_fa.logits_per_image))}",
```
https://github.com/huggingface/transformers/pull/31499/commits/598792f8ee57d6daade6263aa91fb225111f14a4

qubvel committed:
Add math sdpa backend for all options

qubvel committed:
[run slow] siglip

qubvel commented on src/transformers/models/siglip/modeling_siglip.py:
```diff
@@ -786,7 +1069,7 @@ def forward(
         # note: SigLIP's text model does not use a causal mask, unlike the original CLIP model.
         # expand attention_mask
-        if attention_mask is not None:
+        if attention_mask is not None and not self._use_flash_attention_2:
```
I guess we don't need it. I additionally tested with the following code and with the same attention mask preparation outputs matched for both implementations `eager` and `sdpa`

```python
import torch

from transformers.modeling_attn_mask_utils import _prepare_4d_attention_mask

from transformers import SiglipConfig
from transformers.models.siglip.modeling_siglip import SiglipAttention, SiglipSdpaAttention

torch.manual_seed(235093093)

dtype = torch.float16
device = "cuda"

# Configure
config = SiglipConfig()

hidden_size = 6
num_attention_heads = 1
seq_len = 5
batch_size = 1

config.vision_config.hidden_size = hidden_size
config.vision_config.num_attention_heads = num_attention_heads

# Eager attention
attention = SiglipAttention(config.vision_config)
attention = attention.to(dtype).to(device)

# SDPA attention
attention_sdpa = SiglipSdpaAttention(config.vision_config)
attention_sdpa.load_state_dict(attention.state_dict())
attention_sdpa = attention_sdpa.to(dtype).to(device)

# Prepare inputs
dummy_input = torch.rand(
    [batch_size, seq_len, hidden_size], dtype=dtype, device=device,
)
dummy_attention_mask = torch.ones(
    [batch_size, seq_len], dtype=dtype, device=device,
)

# padding
dummy_attention_mask[:1, -2:] = 0
print("Dummy attention mask:\n", dummy_attention_mask)

# Prepare attention mask
dummy_attention_mask_eager = _prepare_4d_attention_mask(
    dummy_attention_mask, dummy_input.dtype
)  # 1, 1, 512, 512  -> batch_size, 1, seq_len, seq_len

# the same for SDPA
dummy_attention_mask_sdpa = dummy_attention_mask_eager

with torch.no_grad():
    attn_output, attn_weights = attention(dummy_input, dummy_attention_mask_eager)
    attn_output_sdpa, attn_weights_sdpa = attention_sdpa(dummy_input, dummy_attention_mask_sdpa)

print("\nEager:\n", attn_output)
print("\nSDPA:\n", attn_output_sdpa)
print("\nDiff:\n", attn_output - attn_output_sdpa)

diff_with_sdpa = torch.abs(attn_output - attn_output_sdpa).max()
print("\nDiff with SDPA:", diff_with_sdpa)
```

```output
Dummy attention mask:
 tensor([[1., 1., 1., 0., 0.]], device='cuda:0', dtype=torch.float16)

Eager:
 tensor([[[ 0.0267,  0.3291,  0.3442,  0.6152, -0.1914,  0.1541],
         [ 0.0174,  0.3289,  0.3398,  0.6138, -0.1968,  0.1560],
         [ 0.0283,  0.3301,  0.3447,  0.6162, -0.1914,  0.1542],
         [ 0.0176,  0.3289,  0.3401,  0.6138, -0.1968,  0.1559],
         [ 0.0228,  0.3293,  0.3423,  0.6147, -0.1941,  0.1550]]],
       device='cuda:0', dtype=torch.float16)

SDPA:
 tensor([[[ 0.0267,  0.3291,  0.3442,  0.6152, -0.1914,  0.1541],
         [ 0.0174,  0.3289,  0.3398,  0.6138, -0.1968,  0.1560],
         [ 0.0283,  0.3301,  0.3447,  0.6162, -0.1914,  0.1542],
         [ 0.0176,  0.3289,  0.3401,  0.6138, -0.1968,  0.1559],
         [ 0.0228,  0.3293,  0.3423,  0.6147, -0.1941,  0.1550]]],
       device='cuda:0', dtype=torch.float16)

Diff:
 tensor([[[0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0.]]], device='cuda:0', dtype=torch.float16)

Diff with SDPA: tensor(0., device='cuda:0', dtype=torch.float16)
```

lucasjinreal commented:
Thanks for the work! 

Which latest version on pypi would support this feature?

qubvel commented:
Hi @lucasjinreal, it's going to be included in the next release, most probably 4.43.0. You can try it now by installing transformers from the source

```bash
pip install -U git+https://github.com/huggingface/transformers.git
```

lucasjinreal commented:
I will waiting for the 4.43, my server unable to access network, any estimate time will 4.43 out?

amyeroberts commented:
@lucasjinreal Just to confirm, you can't access github from your server? If it's just relating to internet, you would also need access for installing the latest release from pypi. 

We typically release on a monthly schedule. You can see the list of [releases here](https://github.com/huggingface/transformers/releases). The next release will probably be in 2-3 weeks.

lucasjinreal commented:
thanks, will try update weeks later.
